src/transformStateWithClones.js

```diff
@@ -7,7 +7,31 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = {};
+
+  Object.assign(currentState, state);
+
+  for (let i = 0; i < actions.length; i++) {
+    const updatedState = {};
+
+    Object.assign(updatedState, currentState);
+
+    if (actions[i].type === 'addProperties') {
+      Object.assign(updatedState, actions[i].extraData);
+    } else if (actions[i].type === 'removeProperties') {
+      actions[i].keysToRemove.forEach((key) => delete updatedState[key]);
+    } else if (actions[i].type === 'clear') {
+      for (const key in updatedState) {
+        delete updatedState[key];
+      }
+    }
+
+    stateHistory.push(updatedState);
+    currentState = updatedState;
```

Comment:
Directly assigning `updatedState` to `currentState` can lead to unintended side effects because `updatedState` is a reference to the same object. Consider using `Object.assign` to create a new object for `currentState` to avoid mutating the original state.